feat: linea docs (#63)
LimeChain/RollupCodes: src/docs/linea.mdx

</Section>

<Section title="General">

    <Parameter name="Block Time" value="12 seconds" tooltip="The rate at which the rollup produces blocks. Keep in mind that the value is subject to change in the future"/>

    <Parameter name="Gas Limit" value="61 million" tooltip="The gas limit that can be consumed by an L2 block"/>

    <Parameter name="Sequencing Frequency" value="30 - 120 seconds" tooltip="The frequency at which the rollup posts L2 transactions on Ethereum L1" />

    <Parameter name="Finality" value="8 - 32 hours" tooltip="The time it takes for an L2 transaction to be considered final after it is sequenced on L1" />

    <Parameter name="Rollup’s State Contract on L1" value={<Reference label="0xb32c3D0dDb0063FfB15E8a50b40cC62230D820B3" url="https://etherscan.io/address/0xb32c3D0dDb0063FfB15E8a50b40cC62230D820B3" />} tooltip="The contract used for sequencing, proving and storing the state of the L2 network" />

    <Parameter name="State Trie" value="Sparse Merkle Trie + MiMC Hashing" tooltip="The data structure used to represent the state of the Rollup along with the hashing algorithm used to compute the root of the trie" />

>     <Parameter name="Node Implementations" value={[<Reference key="0" label="geth (Go)" url="https://github.com/Consensys/go-ethereum" />, <Reference key="1" label="linea-besu (Work in progress)" url="https://github.com/Consensys/linea-besu" />]} tooltip="The different client implementations of the Rollup" />

Reviewer: This is not the node that is running the Linea rollup IMO. It is a simple geth fork without any changes.

---
title: "Linea"

> subtitle: "Linea is a ZK-Rollup that targets EVM compatibility. The protocol team centrally operates a closed-source coordinator (sequencer + prover). Due to its affiliation with ConsenSys, it is positioned to potentially benefit from better utilisation of the ecosystem's tools and infrastructure. The Rollup uses quantum-resistant cryptography for its validity proofs."

Reviewer: ```suggestion
subtitle: "Linea is a ZK-Rollup that targets EVM compatibility. The protocol team centrally operates a closed-source coordinator (sequencer + prover). As it is a product of ConsenSys, it is positioned to potentially benefit from better utilisation of the ecosystem's tools and infrastructure. The Rollup uses quantum-resistant cryptography for its validity proofs."
```

labels:
  - ZK Rollup
  - EVM
  - Mainnet
links:
  website:
    url: "https://linea.build/"
    name: "linea.build"
  docs:
    url: "https://docs.linea.build/"
    name: "Docs"
  l2beat:
    url: "https://l2beat.com/scaling/projects/linea"
    name: "L2Beat"
  github:
    url: "https://github.com/Consensys/linea-contracts"
    name: "Github"
  twitter:
    url: "https://twitter.com/LineaBuild"
    name: "LineaBuild"
  discord:
    url: "https://discord.com/invite/linea"
    name: "Discord"
---

<Section title="Overview">

> Linea is a ZK-Rollup that targets EVM compatibility. The protocol team centrally operates a closed-source coordinator (sequencer + prover). Due to its affiliation with ConsenSys, it is positioned to potentially benefit from better utilisation of the ecosystem's tools and infrastructure. The Rollup uses quantum-resistant cryptography for its validity proofs.

Reviewer: ```suggestion
Linea is a ZK-Rollup that targets EVM compatibility. The protocol team centrally operates a closed-source coordinator (sequencer + prover). As it is a product of ConsenSys, it is positioned to potentially benefit from better utilisation of the ecosystem's tools and infrastructure. The Rollup uses quantum-resistant cryptography for its validity proofs.
```

</Section>

<Section title="General">

    <Parameter name="Block Time" value="12 seconds" tooltip="The rate at which the rollup produces blocks. Keep in mind that the value is subject to change in the future"/>

    <Parameter name="Gas Limit" value="61 million" tooltip="The gas limit that can be consumed by an L2 block"/>

    <Parameter name="Sequencing Frequency" value="30 - 120 seconds" tooltip="The frequency at which the rollup posts L2 transactions on Ethereum L1." />

    <Parameter name="Finality" value="8 - 32 hours" tooltip="The time it takes for validity proof to be generated and submitted which makes an L2 transaction final. ZK proof generation takes around 15 minutes, but its submission is delayed by at least 8 hours as a security measure during the alpha phase. The protocol may delay the submission up to 32 hours in order to minimize the L1 gas cost." />

    <Parameter name="Rollup’s State Contract on L1" value={<Reference label="0xd19d4B5d358258f05D7B411E21A1460D11B0876F" url="https://etherscan.io/address/0xd19d4B5d358258f05D7B411E21A1460D11B0876F" />} tooltip="The contract used for sequencing, proving and storing the state of the L2 network" />

>     <Parameter name="State Trie" value="Sparse Merkle Trie + MiMC Hashing" tooltip="The data structure used to represent the state of the Rollup along with the hashing algorithm used to compute the root of the trie" />

Reviewer: ```suggestion
    <Parameter name="State Trie" value="Main trie is Sparse Merkle Trie + MiMC Hashing. Secondary MPT + Keccak Hashing is maintained for follower nodes" tooltip="The data structure used to represent the state of the Rollup along with the hashing algorithm used to compute the root of the trie" />
```

>     <Parameter name="Node Implementations" value={["The production Sequencer and Prover are closed-source. Geth node supports both SMT + MiMC for state root calculation and MPT + Keccak. The following can be used as follower nodes:", <Reference key="0" label="geth (Go)" url="https://github.com/Consensys/go-ethereum" />, <Reference key="1" label="linea-besu (Work in progress)" url="https://github.com/Consensys/linea-besu" />]} tooltip="The different client implementations of the Rollup" />

Reviewer: ```suggestion
    <Parameter name="Node Implementations" value={["The production Sequencer and Prover are closed-source. The following can be used as follower nodes:", <Reference key="0" label="geth (Go)" url="https://github.com/Consensys/go-ethereum" />, <Reference key="1" label="linea-besu (Work in progress)" url="https://github.com/Consensys/linea-besu" />]} tooltip="The different client implementations of the Rollup" />
```

    | <Copy value="0x508Ca82Df566dCD1B0DE8296e70a96332cD644ec" label="0x508...ec" /> | <Reference label="L2 Message Service" url="https://github.com/Consensys/linea-contracts/blob/main/contracts/messageService/l2/L2MessageService.sol" /> | Provides functionality for sending and receiving messages between domains (Ethereum and Linea zkEVM). |

</Section>

<Section title="RPC-API">

    <Legend />

    The endpoints `eth_gasPrice` and `eth_feeHistory` have [modified underlying logic](https://docs.linea.build/build-on-linea/gas-fees). Base fee is fixed to constant "7".

    The following endpoints behave differently compared to the canonical Ethereum L1 implementation of the JSON RPC API.

    | Method | Params | Rollup Behaviour | Ethereum L1 Behaviour |
    | :----- | :----- | :--------------- | :-------------------- |

>     | `eth_newBlockFilter` | None | Implemented in Node, not available in providers yet  | Creates a filter in the node, to notify when a new block arrives. <Unsupported /> |

Reviewer: ```suggestion
    | `eth_newBlockFilter` | None | Not supported by public nodes. Supported when running your own follower node  | Creates a filter in the node, to notify when a new block arrives. <Unsupported /> |
```

>     | `eth_newFilter` | None | Implemented in Node, not available in providers yet  | Creates a filter object, based on filter options, to notify when the state changes (logs). <Unsupported /> |

Reviewer: ```suggestion
    | `eth_newFilter` | None | Not supported by public nodes. Supported when running your own follower node  | Creates a filter object, based on filter options, to notify when the state changes (logs). <Unsupported /> |
```

    <Parameter name="State Trie" value="Sparse Merkle Trie + MiMC Hashing" tooltip="The data structure used to represent the state of the Rollup along with the hashing algorithm used to compute the root of the trie" />

    <Parameter name="Node Implementations" value={"Geth is the primary node implementation. The production Sequencer and Prover are closed-source, linea-besu is an open-sourced node under development."} tooltip="The different client implementations of the Rollup" />

    <Parameter name="Transaction Types" tooltip="The types of transactions supported on the Rollup" />

        - **Type 0** - User Transaction. Represent [pre-EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) transactions
        - **Type 1** - User Transaction. Represent legacy transactions post [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) transactions
        - **Type 2** - User Transaction. [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) transactions
    ___

>     <MultiRowParameters title="Messaging" tooltip="The Rollup provides native communication with L1. Anyone is able to transmit arbitrary messages and the Arbitrum protocol will provide proof of the data transfer on the destination chain" data={[

Reviewer: ```suggestion
    <MultiRowParameters title="Messaging" tooltip="The Rollup provides native communication with L1. Anyone is able to transmit arbitrary messages and the Linea protocol will provide proof of the data transfer on the destination chain" data={[
```

        {
            title: 'L1 → L2 (automatic)',
            rows: [
                {
                    label: 'Latency',
                    tooltip: 'The time it takes for a message to be made available on the Rollup after it is included in the Ethereum block.',
                    value: '17min'
                },
                {
                    label: 'Cost',
                    tooltip: 'The gas cost of purely sending a message from Ethereum to Linea. Senders on L1 pay an additional fee for the execution of the message on L2.',

>                     value: '66 000 L1 gas'

Reviewer: What is the "additional fee" that is paid for the execution of the message on L2?

                },
                {
                    label: 'Delivery method',
                    tooltip: 'How the smart contract on the target chain is called. In both automatic and manual claiming we are waiting for a central unit to add the message hashes on the target chain (~17 minutes)',
                    value: 'Automatic'
                },
            ]
        },
        {
            title: 'L1 → L2 (manual)',
            rows: [
                {
                    label: 'Latency',
                    tooltip: 'The time it takes for a message to be made available on the Rollup after it is included in the Ethereum block.',

>                     value: '> 17 min'

Reviewer: ```suggestion
                    value: '17 min'
```

    | Opcode | Name | Solidity Equivalent | Rollup Behaviour | Ethereum L1 Behaviour |
    | :----- | :--- | :-------------------| :--------------- | :-------------------- |
    | 44 | PREVRANDAO (legacy DIFFICULTY) | `block.difficulty` | Returns a fixed number "2" | Get the output of the randomness beacon provided by the beacon chain <Modified /> |
    | 5F | PUSH0 |  | Not supported yet! <br /><br /> Solidity version `0.8.20` or higher can only be used with an evm version lower than the default `shanghai`. <br /><br /> Versions up to `0.8.19` (included) are fully compatible | Places value 0 on the stack <Unsupported /> |

</Section>

<Section title="Precompiled Contracts">

    > ⚠️ All Precompiled Contacts defined in the canonical Ethereum L1 implementation (until London hard fork) have the same behaviour on the Rollup, however, the execution of those Precompiles is not covered by the circuits and therefore not present in the Validity proof

</Section>

<Section title="System Contracts">

    The Rollup has introduced the following system contracts.

    | Address | Name | Description |
    | :--- | :--- | :--- |

>     | <Copy value="0xC84832f69bFFbC1A94E44a157A342766E959ED27" label="0xC84...27" /> | <Reference label="PlonkVerifierFull" url="https://github.com/Consensys/linea-contracts/blob/main/contracts/verifiers/PlonkVerifierFull.sol" /> | Plonk verifier contract used by the Linea zkEVM rollup. |

Reviewer: When we refer to "system contracts" in other rollups we are listing only the contracts that end users/developers can utilise. Think of it as "additional features" that are provided by the rollup except for the precompile and opcode functionality. 

Having that in mind, I do not think that the Plonk Verifier should be here.

>     | <Copy value="0xb32c3D0dDb0063FfB15E8a50b40cC62230D820B3" label="0xb32...B3" /> | <Reference label="ZkEvmV2" url="https://github.com/Consensys/linea-contracts/blob/main/contracts/ZkEvmV2.sol" /> |The main contract of the Linea zkEVM rollup. Contains state roots, the verifier addresses and manages messages between L1 and the L2.|

Reviewer: This is the "implementation" contract for the zkEVMv2. The proxy contract is `0xd19d4B5d358258f05D7B411E21A1460D11B0876F`. We can leave the `address` value to the proxy contract as this is the entrypoint and in the `Description` state:

```The main contract of the Linea zkEVM Rollup. Contains the L2 state roots. It handles sequencing and finalisation of L2 blocks and provides messaging functionality between L1 and L2. Provides functionality for sending and receiving messages between domains (Ethereum and Linea). Details on the [L1 Message Service](https://github.com/Consensys/linea-contracts/blob/main/contracts/messageService/l1/L1MessageService.sol)```

>     | <Copy value="0xd19d4B5d358258f05D7B411E21A1460D11B0876F" label="0xd19...6F" /> | <Reference label="L1 Message Service" url="https://github.com/Consensys/linea-contracts/blob/main/contracts/messageService/l1/L1MessageService.sol" /> | L1 Message Service.|

Reviewer: We can remove this line as we have effectively described it in the above row

>     | <Copy value="0x508Ca82Df566dCD1B0DE8296e70a96332cD644ec" label="0x508...ec" /> | <Reference label="L2 Message Service" url="https://github.com/Consensys/linea-contracts/blob/main/contracts/messageService/l2/L2MessageService.sol" /> | L2 Message Service.|

Reviewer: We can provide the following description:

```Provides functionality for sending and receiving messages between domains (Ethereum and Linea zkEVM)```

    | Method | Params | Rollup Behaviour | Ethereum L1 Behaviour |
    | :----- | :----- | :--------------- | :-------------------- |
    | `eth_newBlockFilter` | None | Not supported  | Creates a filter in the node, to notify when a new block arrives. <Unsupported /> |
    | `eth_newFilter` | None | Not supported  | Creates a filter object, based on filter options, to notify when the state changes (logs). <Unsupported /> |
    
</Section>

<Section title="Transaction Fees">

    Other than the L1 gas fee component, Linea charges L2 gas fees which should be around 1/15th of Ethereum's.

    [L2 Gas fees](https://docs.linea.build/build-on-linea/gas-fees).

</Section>

<Section title="Tools">

    Existing EVM-based tooling is supported such as ethers, web3.js, ethers-rs, hardhat, foundry and truffle.

    [Linea SDK](https://docs.linea.build/build-on-linea/linea-sdk) focuses on interacting with smart contracts on both Ethereum and Linea networks and provides custom functions to obtain message information.

    [Linea Safe](https://safe.linea.build/welcome) is a white-label fork of Gnosis Safe deployed by the Linea team.

>     [Ape framework](https://github.com/Consensys/ape-linea) is a plugin to add support to the Ape cli web3 development tool for Linea developed by the Linea team.

Reviewer: ```suggestion
    [Ape framework](https://github.com/Consensys/ape-linea) is a Web3 Development framework in Python developed by the Linea team.
```

---
title: "Linea"
subtitle: "Linea is a ZK-Rollup powered by Consensys zkEVM. The protocol team centrally operates the coordinator (sequencer + prover) which is closed source. They are utilizing quantum-resistant cryptography for their zk-proofs. The rollup has rich tooling and infrastructure support."
labels:
  - ZK Rollup
  - EVM
  - Mainnet
links:
  website:
    url: "https://linea.build/"
    name: "linea.build"
  docs:
    url: "https://docs.linea.build/"
    name: "Docs"
  l2beat:
    url: "https://l2beat.com/scaling/projects/linea"
    name: "L2Beat"
  github:
    url: "https://github.com/Consensys/linea-contracts"
    name: "Github"
  twitter:
    url: "https://twitter.com/LineaBuild"
    name: "LineaBuild"
  discord:
    url: "https://discord.com/invite/linea"
    name: "Discord"
---

<Section title="Overview">
Linea is a ZK-Rollup powered by Consensys zkEVM. The protocol team centrally operates the coordinator (sequencer + prover) which is closed source. They are utilizing quantum-resistant cryptography for their zk-proofs. The rollup has rich tooling and infrastructure support.

###### Focus
- Open sourcing rollup operators

> - Support for multiple zk-proof algorithms

Author: https://docs.linea.build/decentralization-roadmap

> - Support for multiple zk-proof algorithms

Reviewer: added a suggestion

    | Opcode | Name | Solidity Equivalent | Rollup Behaviour | Ethereum L1 Behaviour |
    | :----- | :--- | :-------------------| :--------------- | :-------------------- |

>     | 44 | PREVRANDAO (legacy DIFFICULTY) | `block.difficulty` | Returns a fixed number | Get the output of the randomness beacon provided by the beacon chain <Modified /> |

Author: 2

    | 5F | PUSH0 |  | Not supported yet! <br /><br /> Solidity version `0.8.20` or higher can only be used with an evm version lower than the default `shanghai`. <br /><br /> Versions up to `0.8.19` (included) are fully compatible | Places value 0 on the stack <Unsupported /> |

> </Section>

Reviewer: If the precompiles are working as in Ethereum L1 but they are not part of the validity proof lets add a "callout" similar to the zkSync warning in the OPCODES section:

"All Precompiled Contacts defined in the canonical Ethereum L1 implementation (until London hard fork) have the same behaviour on the Rollup, however, the execution of those Precompiles is not covered by the circuits and therefore not present in the Validity proof"

Reviewer: Based on the answer that we receive from the TG group we can list the specific precompiles that are covered on mainnet not - if there are such

    Existing EVM-based tooling is supported such as ethers, web3.js, ethers-rs, hardhat, foundry and truffle.

    [Linea SDK](https://docs.linea.build/build-on-linea/linea-sdk) focuses on interacting with smart contracts on both Ethereum and Linea networks and provides custom functions to obtain message information.

>     [Linea Safe](https://safe.linea.build/welcome) is a multisig wallet for Linea.

Reviewer: let's change it to:

[Linea Safe](https://safe.linea.build/welcome) is a white-label fork of Gnosis Safe deployed by the Linea team.

>     [Ape framework](https://github.com/Consensys/ape-linea) is a plugin developed by the Linea team.

Reviewer: What does it do?

Author: adds support to the Ape cli web3 development tool for Linea

Author: python tool giving similar functionality to foundry
